streaming/in.go

```diff
@@ -296,20 +295,22 @@ func (c *crypt4GHInternalReader) read(p []byte) (n int, err error) {
 				}
 				haveRead++
 			}
-		} else {
-			// Read larger chunk from buffer. As precaution, limit to what we
-			// should be able to read only, as that is the bit we've checked
-			// if the discard list imposes any holes in
-
+		case canRead > 0:
 			r, err := c.buffer.Read(p[haveRead : haveRead+canRead])
 			haveRead += r
-			c.streamPos += int64(r)
+			if err != nil {
+				return haveRead, err
+			}
 
-			// Not sure why we'd get an error here, but forward
-			// if we see that
+			c.streamPos += int64(r)
```

Comment:
Nothing in our test suite, (including the ones added in #107), indicates that there will be an issue, any calling function should trigger on the error and then it's logical to restart the read from the last reported OK position.

Reply:
No, the tests aren't designed to trigger for such an issue (and the code as it is only reads the data available so won't see EOF there, again why this fix is needed). Giving that we're working with a buffer, I believe we're unlikely to see any errors in practice.

Philosophically though, I believe this code is wrong - just like one risk missing data if checking/acting on an error rather than looking at the amount of data read first.

Assuming we'd see an actual (recoverable) error from this `Read`, we'd risk handing out the same data twice, corrupting the data stream the client gets.